File: backend/src/models/Agent.ts
import { db } from '../services/db.service';
import { Agent, AgentType, AgentStatus } from '../types';
import { v4 as uuidv4 } from 'uuid';

export class AgentModel {
    /**
     * Get all agents
     */
    static async findAll(): Promise<Agent[]> {
        const result = await db.query(`
      SELECT * FROM agents 
      ORDER BY created_at DESC
    `);
        return result.rows;
    }

    /**
     * Get agent by ID
     */
    static async findById(id: string): Promise<Agent | null> {
        const result = await db.query(
            'SELECT * FROM agents WHERE id = $1',
            [id]
        );
        return result.rows[0] || null;
    }

    /**
     * Get agent by type
     */
    static async findByType(type: AgentType): Promise<Agent | null> {
        const result = await db.query(
            'SELECT * FROM agents WHERE type = $1',
            [type]
        );
        return result.rows[0] || null;
    }

    /**
     * Create new agent
     */
    static async create(data: Omit<Agent, 'id' | 'createdAt' | 'updatedAt'>): Promise<Agent> {
        const id = uuidv4();
        const result = await db.query(
            `INSERT INTO agents (id, type, name, description, status, config)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
            [id, data.type, data.name, data.description, data.status, JSON.stringify(data.config)]
        );
        return result.rows[0];
    }

    /**
     * Update agent
     */
    static async update(id: string, data: Partial<Agent>): Promise<Agent | null> {
        const updates: string[] = [];
        const values: any[] = [];
        let paramCount = 1;

        if (data.name !== undefined) {
            updates.push(`name = $${paramCount++}`);
            values.push(data.name);
        }
        if (data.description !== undefined) {
            updates.push(`description = $${paramCount++}`);
            values.push(data.description);
        }
        if (data.status !== undefined) {
            updates.push(`status = $${paramCount++}`);
            values.push(data.status);
        }
        if (data.config !== undefined) {
            updates.push(`config = $${paramCount++}`);
            values.push(JSON.stringify(data.config));
        }

        if (updates.length === 0) {
            return await this.findById(id);
        }

        updates.push(`updated_at = NOW()`);
        values.push(id);

        const result = await db.query(
            `UPDATE agents SET ${updates.join(', ')} WHERE id = $${paramCount} RETURNING *`,
            values
        );

        return result.rows[0] || null;
    }

    /**
     * Delete agent
     */
    static async delete(id: string): Promise<boolean> {
        const result = await db.query('DELETE FROM agents WHERE id = $1', [id]);
        return (result.rowCount || 0) > 0;
    }
}

File: backend/internal/models/agent.go
package models

import (
	"database/sql"
	"encoding/json"
	"time"

	"go-backend/internal/database"
)

type AgentStatus string

const (
	AgentStatusActive AgentStatus = "active"
	AgentStatusPaused AgentStatus = "paused"
	AgentStatusError  AgentStatus = "error"
)

type Agent struct {
	ID          string          `json:"id"`
	Type        string          `json:"type"`
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Status      AgentStatus     `json:"status"`
	Config      json.RawMessage `json:"config"`
	CreatedAt   time.Time       `json:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at"`
}

type AgentRepository struct{}

func NewAgentRepository() *AgentRepository {
	return &AgentRepository{}
}

func (r *AgentRepository) FindAll() ([]Agent, error) {
	query := `SELECT id, type, name, description, status, config, created_at, updated_at FROM agents ORDER BY created_at DESC`

	rows, err := database.DB.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	agents := []Agent{}
	for rows.Next() {
		agent := Agent{}
		var config sql.NullString

		err := rows.Scan(
			&agent.ID, &agent.Type, &agent.Name, &agent.Description,
			&agent.Status, &config, &agent.CreatedAt, &agent.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		if config.Valid {
			agent.Config = json.RawMessage(config.String)
		} else {
			agent.Config = json.RawMessage("{}")
		}

		agents = append(agents, agent)
	}

	return agents, nil
}

func (r *AgentRepository) FindByID(id string) (*Agent, error) {
	query := `SELECT id, type, name, description, status, config, created_at, updated_at FROM agents WHERE id = $1`

	agent := &Agent{}
	var config sql.NullString

	err := database.DB.QueryRow(query, id).Scan(
		&agent.ID, &agent.Type, &agent.Name, &agent.Description,
		&agent.Status, &config, &agent.CreatedAt, &agent.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	if config.Valid {
		agent.Config = json.RawMessage(config.String)
	} else {
		agent.Config = json.RawMessage("{}")
	}

	return agent, nil
}

func (r *AgentRepository) FindByType(agentType string) (*Agent, error) {
	query := `SELECT id, type, name, description, status, config, created_at, updated_at FROM agents WHERE type = $1`

	agent := &Agent{}
	var config sql.NullString

	err := database.DB.QueryRow(query, agentType).Scan(
		&agent.ID, &agent.Type, &agent.Name, &agent.Description,
		&agent.Status, &config, &agent.CreatedAt, &agent.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	if config.Valid {
		agent.Config = json.RawMessage(config.String)
	} else {
		agent.Config = json.RawMessage("{}")
	}

	return agent, nil
}

func (r *AgentRepository) Update(id string, updates map[string]any) (*Agent, error) {
	// Build dynamic update query
	setClauses := []string{}
	args := []any{}
	argNum := 1

	if name, ok := updates["name"].(string); ok {
		setClauses = append(setClauses, "name = $"+string(rune('0'+argNum)))
		args = append(args, name)
		argNum++
	}
	if description, ok := updates["description"].(string); ok {
		setClauses = append(setClauses, "description = $"+string(rune('0'+argNum)))
		args = append(args, description)
		argNum++
	}
	if status, ok := updates["status"].(string); ok {
		setClauses = append(setClauses, "status = $"+string(rune('0'+argNum)))
		args = append(args, status)
		argNum++
	}
	if config, ok := updates["config"]; ok {
		configJSON, _ := json.Marshal(config)
		setClauses = append(setClauses, "config = $"+string(rune('0'+argNum)))
		args = append(args, configJSON)
		argNum++
	}

	if len(setClauses) == 0 {
		return r.FindByID(id)
	}

	// Always update updated_at
	setClauses = append(setClauses, "updated_at = NOW()")
	args = append(args, id)

	query := "UPDATE agents SET "
	for i, clause := range setClauses {
		if i > 0 {
			query += ", "
		}
		query += clause
	}
	query += " WHERE id = $" + string(rune('0'+argNum)) + " RETURNING id, type, name, description, status, config, created_at, updated_at"

	agent := &Agent{}
	var config sql.NullString

	err := database.DB.QueryRow(query, args...).Scan(
		&agent.ID, &agent.Type, &agent.Name, &agent.Description,
		&agent.Status, &config, &agent.CreatedAt, &agent.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	if config.Valid {
		agent.Config = json.RawMessage(config.String)
	} else {
		agent.Config = json.RawMessage("{}")
	}

	return agent, nil
}
